courseware/Preconfigured_EBM.md
---
jupytext:
  formats: ipynb,md:myst,py:percent
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.12
    jupytext_version: 1.8.0
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

# Preconfigured Energy Balance Models

+++

In this document the basic use of climlab's preconfigured EBM class is shown. 

Contents are how to

  * setup an EBM model
  * show and access subprocesses
  * integrate the model
  * access and plot various model variables
  * calculate the global mean of the temperature

```{code-cell}
from __future__ import division, print_function
import numpy as np
import matplotlib.pyplot as plt
import climlab
from climlab import constants as const
```

### Model Creation

+++

The regular path for the EBM class is ``climlab.model.ebm.EBM`` but it can also be accessed through ``climlab.EBM``

An EBM model instance is created through

```{code-cell}
# model creation
ebm_model = climlab.EBM()
```

By default many parameters are set during initialization:

``num_lat=90, S0=const.S0, A=210., B=2., D=0.55, water_depth=10., Tf=-10, a0=0.3, a2=0.078, ai=0.62, timestep=const.seconds_per_year/90., T0=12., T2=-40``

For further details see the climlab documentation.

Many of the input parameters are stored in the following dictionary:

```{code-cell}
# print model parameters
ebm_model.param
```

The model consists of one state variable (surface temperature) and a couple of defined subprocesses.

```{code-cell}
# print model states and suprocesses
print(ebm_model)
```

### Model subprocesses

+++

The subprocesses are stored in a dictionary and can be accessed through

```{code-cell}
# access model subprocesses
ebm_model.subprocess.keys()
```

So to access the time type of the Longwave Radiation subprocess for example, type:

```{code-cell}
# access specific subprocess through dictionary
ebm_model.subprocess['LW'].time_type
```

```{code-cell}
#  For interactive convenience, you can also use attribute access for the same thing:
ebm_model.subprocess.LW.time_type
```

### Model integration

+++

The model time dictionary shows information about all the time related content and quantities.

```{code-cell}
# accessing the model time dictionary
ebm_model.time
```

To integrate the model forward in time different methods are availible:

```{code-cell}
# integrate model for a single timestep
ebm_model.step_forward()
```

The model time step has increased from 0 to 1:

```{code-cell}
ebm_model.time['steps']
```

```{code-cell}
# integrate model for a 50 days
ebm_model.integrate_days(50.)
```

```{code-cell}
# integrate model for two years
ebm_model.integrate_years(1.)
```

```{code-cell}
# integrate model until solution converges
ebm_model.integrate_converge()
```

## Plotting model variables

+++

A couple of interesting model variables are stored in a dictionary named ``diagnostics``. It has following entries:

```{code-cell}
ebm_model.diagnostics.keys()
```

They can be accessed in two ways:

- Through dictionary methods like ``ebm_model.diagnostics['ASR']``
- As process attributes like ``ebm_model.ASR``

```{code-cell}
ebm_model.icelat
```

The following code does the plotting for some model variables.

```{code-cell}
# creating plot figure
fig = plt.figure(figsize=(15,10))

# Temperature plot
ax1 = fig.add_subplot(221)
ax1.plot(ebm_model.lat,ebm_model.Ts)

ax1.set_xticks([-90,-60,-30,0,30,60,90])
ax1.set_xlim([-90,90])
ax1.set_title('Surface Temperature', fontsize=14)
ax1.set_ylabel('(degC)', fontsize=12)
ax1.grid()

# Albedo plot
ax2 = fig.add_subplot(223, sharex = ax1)
ax2.plot(ebm_model.lat,ebm_model.albedo)

ax2.set_title('Albedo', fontsize=14)
ax2.set_xlabel('latitude', fontsize=10)
ax2.set_ylim([0,1])
ax2.grid()

# Net Radiation plot
ax3 = fig.add_subplot(222, sharex = ax1)
ax3.plot(ebm_model.lat, ebm_model.OLR, label='OLR',
                                       color='cyan')
ax3.plot(ebm_model.lat, ebm_model.ASR, label='ASR',
                                       color='magenta')
ax3.plot(ebm_model.lat, ebm_model.ASR-ebm_model.OLR, 
                                       label='net radiation',
                                       color='red')

ax3.set_title('Net Radiation', fontsize=14)
ax3.set_ylabel('(W/m$^2$)', fontsize=12)
ax3.legend(loc='best')
ax3.grid()

# Energy Balance plot
net_rad = np.squeeze(ebm_model.net_radiation)
transport = ebm_model.heat_transport_convergence()

ax4 = fig.add_subplot(224, sharex = ax1)
ax4.plot(ebm_model.lat, net_rad, label='net radiation', 
                                              color='red')
ax4.plot(ebm_model.lat, transport, label='heat transport', 
                                              color='blue')
ax4.plot(ebm_model.lat, net_rad+transport, label='balance',
                                             color='black')

ax4.set_title('Energy', fontsize=14)
ax4.set_xlabel('latitude', fontsize=10)
ax4.set_ylabel('(W/m$^2$)', fontsize=12)
ax4.legend(loc='best')
ax4.grid()


plt.show()
```

The energy balance is zero at every latitude. That means the model is in equilibrium. Perfect!

+++

### Global mean temperature

+++

The model's state dictionary has following entries:

```{code-cell}
ebm_model.state.keys()
```

Like diagnostics, state variables can be accessed in two ways:

- With dictionary methods, ``ebm_model.state['Ts']`` 
- As process attributes, ``ebm_model.Ts``

These are entirely equivalent:

```{code-cell}
ebm_model.Ts is ebm_model.state['Ts']
```

The global mean of the model's surface temperature can be calculated through

```{code-cell}
print('The global mean temperature is %.2f deg C.' %climlab.global_mean(ebm_model.Ts))
print('The modeled ice edge is at %.2f deg latitude.' %np.max(ebm_model.icelat))
```

```{code-cell}

```


climlab/radiation/cam3rad/parameters.py
import math

class Parameters:
    '''
    '''
    def __init__(self,**kwargs):

        # Set default param values
        self._setDefaults()

        # Used items will be popped from kwargs
        self.unused_kwargs = kwargs

        # Update params requested in kwargs
        for key in self.value:
            if key in kwargs and kwargs[key] is not None:
                self.value[key] = kwargs.pop(key)

    def _setDefaults(self):
        self.value = {}
        self.units = {}
        self.long_name = {}
        
        # Physical constants
        self.append('Cpd',     1005.7,  'J K-1 kg-1', 'Spec. heat dry air at const. press.') 
        self.append('Cvd',     718.,    'J K-1 kg-1', 'Spec. heat dry air at const. volume') 
        self.append('Cpv',     1870.,   'J K-1 kg-1', 'Specific heat of water vapour at const. press') 
        self.append('Cl',      4190.,   'J K-1 kg-1', 'Specific heat pure liquid water')
        self.append('Csw',     3985.,   'J K-1 kg-1', 'Specific heat sea water')
        self.append('Rv',      461.5,   'J K-1 kg-1', 'Gas constant for water vapour')
        self.append('Rd',      287.04,  'J K-1 kg-1', 'Gas constant dry air')
        self.append('Lv',      2.501e6,          ' ', 'Latent heat vaporisation')
        self.append('Lf',      3.336e5,          ' ', 'Lat heat of fusion')
        self.append('rowl',    1000.,       'kg m-3', 'Density of liquid water')
        self.append('stebol',  5.67e-8,          ' ', 'Stefan Boltzmann constant' )
        self.append('epsilon', 287.04/461.5,     '-', 'Rd/Rv')

        # Orbital and planetary parameters
        self.append('eccen',          0.016715,       '-', 'Eccentricity of orbit')
        self.append('obliq',           23.4441,     'deg', 'Planetary obliquity')
        self.append('prece',             102.7,     'deg', 'Precession of vernal equinox')
        self.append('orb_year',           1995,       '-', 'Earth year for which orbital params computed')
        self.append('lod',              86400.,       's', 'Length of day' ) 
        self.append('daysperyear',      365.25,       '-', 'Number of days in a year' )
        self.append('calday',             80.5,    'days', 'Julian calendar day' )
        self.append('r',              6374000.,       'm', 'Planetary radius' ) 
        self.append('g',                   9.8,   'm s-2', 'Gravitational acceleration')
        self.append('omega', 2.*math.pi/86400., 'rad s-1', 'Planetary rotation rate' ) 
        self.append('radius',               1.,      'AU', 'Mean orbital radius' )
        self.append('scon',              1367.,   'W m-2', 'Solar irradiance at mean orbital radius')

        # Parameters for axisymmetric dynamics
        self.append('Newt', 1.e9, 'days', 'Newtonian cooling timescale' )
        self.append('delh', 100.,    'K', 'Eq-pole gradient of target temp' )
        self.append('delv', 40.,     'K', 'Surface-trop pot temp increase' )

        # Parameters for simple turbulence
        self.append('nuv',             1., 'm^2/s',  'Kinematic viscosity' )
        self.append('Pr',              1., '-',      'Prandtl number (visc/cond)' )
        self.append('do_srf_mom_flx',   1, '-',      'Momentum flux thru surface: 1=Yes, 0=No' )
        self.append('do_srf_sen_flx',   1, '-',      'Heat flux thru surface: 1=Yes, 0=No' )
        self.append('do_srf_lat_flx',   1, '-',      'Moist flux thru surface: 1=Yes, 0=No' )
        self.append('Cd',          1.3e-3, '-',      'Surface flux coeff' )
        self.append('u0',              5., 'm s-1',  'surface wind for bulk surface flux' )

        # Parameters for radiative transfer
        self.append('do_sw',   1,     '-',      '1=do, 0=do not compute SW' )
        self.append('do_lw',   1,     '-',      '1=do, 0=do not compute LW' )
        self.append('in_cld',  0,     '-',      '1=in-cloud, 0=grid avg cloud water path' )
        self.append('co2',     380.,  'ppmv',   'CO2 conc.' )
        self.append('n2o',     1.e-9, 'ppmv',   'N2O conc.' )
        self.append('ch4',     1.e-9, 'ppmv',   'CH4 conc.' )
        self.append('cfc11',   1.e-9, 'ppmv',   'CFC11 conc.' )
        self.append('cfc12',   1.e-9, 'ppmv',   'CFC12 conc.' )
        self.append('cfc22',   1.e-9, 'ppmv',   'CFC22 conc.' ) 
        self.append('o2', .2095, '-', 'O2 volumetric mixing ratio')
        self.append('ccl4', 0., '-', 'volume mixing ratio')
        self.append('tauaer_sw', 0., '-', 'Aerosol optical depth')
        self.append('ssaaer_sw', 0., '-', 'Aerosol single scattering albedo')
        self.append('asmaer_sw', 0., '-', 'Aerosol asymmetry parameter')
        self.append('tauaer_lw', 0., '-', 'Aerosol optical depth')
        self.append('lw_surface_emissivity', 1., '-', 'lw surface emissivity')
        self.append('tauvis',  0.,    '-',      'Aerosol opt. depth, CCM3 scheme' )    
        self.append('tau_inf', 1.,    '-',      'Total optical depth, greygas scheme')
        self.append('alpha_greygas', 1., '-',    'tau profile shape parameter, greygas scheme' )
        self.append('beta_greygas', 1., '-',    'bandwith, greygas scheme' )

        # Parameters for Emanuel's convection scheme
        self.append('elcrit',   0.0011,   '-',      ' ' )
        self.append('tlcrit',   -55.,     '-',      ' ' )
        self.append('entp',     1.5,      '-',      ' ' )
        self.append('sigd',     0.05,     '-',      ' ' )
        self.append('sigs',     0.12,     '-',      ' ' )
        self.append('omtrain',  50.0,     '-',      ' ' )
        self.append('omtsnow',  5.5,     '-',      ' ' )
        self.append('coeffr',   1.0,     '-',      ' ' )
        self.append('coeffs',   0.8,     '-',      ' ' )
        self.append('cu',       0.7,     '-',      ' ' )
        self.append('beta',     10.0,     '-',      ' ' )
        self.append('dtmax',    0.9,     '-',      ' ' )
        self.append('alpha',    0.2,     '-',      ' ' )
        self.append('damp',     0.1,     '-',      ' ' )

        # Parameters for hard convective adjustment scheme
        self.append('lapse',   9.8/1005.7,  'K m-1',    'Specified lapse rate')

        # Parameters for simplified Betts-Miller convection scheme
        self.append('tau_bm',  7200.,    's',   'B-M relaxation timescale')
        self.append('es0',      611.,  'kPa',   'reference pressure for C-C relation')
        self.append('T00',    273.15,    'K',   'reference temperature for C-C relation')
        self.append('rhbm',      0.8,    '-',   'relative humidity of relaxation')
        self.append('Mv',        18.,  'AMU',   'molecular weight of condensible')
        self.append('Ma',        29.,  'AMU',   'mean molecular weight of dry air')
        
        # Parameters for ocean
        self.append('Hslab',   50.,      'm',  'Depth of slab ocean' )

        # Parameters for timestepping
        self.append('dt',   60*15.,    's',  'Time step' )
        self.append('afc',    0.05,    '-',  'Asselin filter coefficient' )

    def append(self,*arg):
        key = arg[0]
        self.value[key] = arg[1]
        self.units[key] = arg[2]
        self.long_name[key] = arg[3]
        
    # The following methods allow a Parameters instance to be treated as a dictionary
    def __getitem__(self,key):
        try: return self.value[key]
        except: raise IndexError,'\n\n CliMT.Params: %s not a known parameter name' % str(key)
        
    def __setitem__(self,key,value):
        self.value[key] = value
        
    def keys(self):
        return self.value.keys()

    def __iter__(self):
        return self.value.__iter__()
